src/Microsoft.AspNet.Routing/Template/TemplateParser.cs

```diff
@@ -322,10 +323,27 @@ private static bool IsSegmentValid(TemplateParserContext context, TemplateSegmen
             for (var i = 0; i < segment.Parts.Count; i++)
             {
                 var part = segment.Parts[i];
+                
                 if (part.IsParameter && part.IsOptional && segment.Parts.Count > 1)
```

Comment:
this particular case seems valid, we need to think about protecting against {foo}/.{bar?} as optional parameter followed by period has special meaning.